packages/core/src/interpreter.ts

```diff
@@ -541,6 +579,13 @@ export class Interpreter<
 
     return this;
   }
+
+  public input(input: Partial<TExtra['input']>): void {
+    this.send(({
```

Comment:
One thing that might be useful conceptually is imagining both `input` and `output` as first-class parts of the XState API. Imagine we have a `model.withRequiredData()` method to represent a typesafe output from the machine. This could make `data` typesafe.

The benefit of having both `input` and `output` is that we can then use XState machines to represent async functions.

You can, of course, represent input as 'context' - but having it be its own property with its own constraints (must be complete, cannot be assigned to) is IMO important.

Reply:
I agree that input/output, as a concept, might be very handy and might increase the type safety of things. 

It would be great to achieve the parity - have the `input` be available on `xstate.init` event and `output` on the `done.invoke.*`.

> You can, of course, represent input as 'context' - but having it be its own property with its own constraints (must be complete, cannot be assigned to) is IMO important.

I'm not sold on this argument though - IMHO those constraints are not that useful because I don't expect people to think in those terms, I don't expect them to choose one (context vs input) based on those. The completeness can be solved at the type level - by typing event payloads and by grouping the keys in context on a property that has no optional members.

Based on that I really think that event+context should always be encouraged over input. I see how the props thing is similar to, observable stream and one could do something like:
```js
obs.subscribe(data => {
  service.input(data)
})
```
but I'm not sure how useful in practice that is. In this context, we can also deal with multiple **implicit** "sources" so the Partial argument swings back to allowing a Partial (I'm still more inclined more to ship without it though to see what happens as it could be very interesting to read community discussions about this).

I just wonder if making **implicit** updates is not a feature that we should stick with and only open up for them when we have strong-enough use cases for them.